storage/src/hash/mod.rs

```diff
@@ -0,0 +1,444 @@
+    let mut hasher = VarBlake2b::new(HASH_LEN)?;
+
+    match inode {
+        Inode::Empty => return Err(HashingError::UnexpectedEmptyInode),
+        Inode::Value(entries) => {
+            // Inode value:
+            //
+            // |   1   |   1  |     n_1      |  ...  |      n_k      |
+            // +-------+------+--------------+-------+---------------+
+            // | \000  |  \n  | prehash(e_1) |  ...  | prehash(e_k)  |
+            //
+            // where n_i = len(prehash(e_i))
+
+            hasher.update(&[0u8]); // type tag
+            hasher.update(&[entries.len() as u8]);
+
+            // Inode value entry:
+            //
+            // |   (LEB128)  |  len(name)   |   1    |   32   |
+            // +-------------+--------------+--------+--------+
+            // | \len(name)  |     name     |  kind  |  hash  |
+
+            for (name, node) in entries {
+                leb128::write::unsigned(&mut hasher, name.len() as u64)?;
+                hasher.update(name.as_bytes());
+                // \000 for nodes, and \001 for contents.
+                match node.node_kind {
+                    NodeKind::Leaf => hasher.update(&[1u8]),
+                    NodeKind::NonLeaf => hasher.update(&[0u8]),
+                };
+                hasher.update(node.entry_hash);
+            }
+        }
+        Inode::Tree {
+            depth,
+            children,
+            pointers,
+        } => {
+            // Inode tree:
+            //
+            // |   1    | (LEB128) |   (LEB128)    |    1   |  33  | ... |  33  |
+            // +--------+----------+---------------+--------+------+-----+------+
+            // |  \001  |  depth   | len(children) |   \k   | s_1  | ... | s_k  |
+
+            hasher.update(&[1u8]); // type tag
+            leb128::write::unsigned(&mut hasher, *depth as u64)?;
+            leb128::write::unsigned(&mut hasher, *children as u64)?;
+            hasher.update(&[pointers.len() as u8]);
+
+            // Inode pointer:
+            //
+            // |    1    |   32   |
+            // +---------+--------+
+            // |  index  |  hash  |
+
+            for (index, hash) in pointers {
+                hasher.update(&[*index]);
+                hasher.update(hash);
+            }
+        }
+    }
+
+    Ok(hasher.finalize_boxed().as_ref().try_into()?)
+}
+
+// Calculates hash of tree
+// uses BLAKE2 binary 256 length hash function
+// hash is calculated as:
+// <number of child nodes (8 bytes)><CHILD NODE>
+// where:
+// - CHILD NODE - <NODE TYPE><length of string (1 byte)><string/path bytes><length of hash (8bytes)><hash bytes>
+// - NODE TYPE - leaf node(0xff0000000000000000) or internal node (0x0000000000000000)
+pub(crate) fn hash_tree(tree: &Tree) -> Result<EntryHash, HashingError> {
+    // If there are >256 entries, we need to partition the tree and hash the resulting inode
+    if tree.len() > 256 {
+        let entries: Vec<(&String, &Node)> = tree.iter().collect();
```

Comment:
@tizoc 
Bruno, this worries me a lots, like you mentioned upper: `IMPORTANT: entries must be sorted in lexicographic order of the name`

what if we change signature: `tree: &Tree` to something like `Iterator + Ord`
in case somebody could change this
`type Tree = im::OrdMap<String, Node>;`

or at least for now please add `IMPORTANT:` comment to this `type Tree` about ordering for `partition_entries`


or explicit sorting here, will add a lots of overhead?
what do you thing any idea?

Reply:
I would rather avoid any explicit sorting, will just add a comment to `Tree`. Note that the sorted requirement has always been there, it has nothing to do with the new code, it is just that one of the Inode variants has the same constraints as the case for which Inodes are not required.

Reply:
@tizoc 
ok, cool, so please jst "add a comment to Tree"

Reply:
We have this already, but it is too vague, will update it with a better description:

```
// Tree must be an ordered structure for consistent hash in hash_tree
// Currently immutable OrdMap is used to allow cloning trees without too much overhead
pub(crate) type Tree = im::OrdMap<String, Node>;
```